File: src/config/policy.ts
/** Explicit installation controls; upstream defaults remain available outside this profile. */
export const localOnly = (): boolean => process.env.PODCLI_LOCAL_ONLY === "1";
export const strictAI = (): boolean => localOnly() || process.env.PODCLI_AI_PROVIDER === "codex-then-claude";

export function requireCloud(): void {
  if (strictAI()) throw new Error("Cloud services are disabled by the Codex → Claude policy.");
}

export function localRequestError(method: string, path: string, body: unknown): string | null {
  if (!localOnly()) return null;
  if (/^\/api\/(download-video|assets\/url|youtube(?:\/|$)|analytics|thumbnail-studio|thumbnail-config|knowledge\/init)/.test(path)
      || /^\/api\/clips\/[^/]+\/(thumbnail|davinci)/.test(path)
      || (path === "/api/settings" && method !== "GET")) {
    return "This feature is disabled in the local installation.";
  }
  const inspect = (value: unknown): boolean => {
    if (Array.isArray(value)) return value.some(inspect);
    if (value && typeof value === "object") return Object.entries(value).some(([key, item]) => {
      if (["url", "video_url", "source_url"].includes(key) && item) return true;
      if (/^(file_path|video_path|audio_path|logo_path|intro_path|outro_path|image_path|videoPath|filePath|logoPath|introPath|outroPath)$/.test(key)
          && typeof item === "string" && (item.includes("://") || item.startsWith("\\\\") || item.startsWith("//"))) return true;
      if (key === "engine" && item && item !== "whisper-py") return true;
      return inspect(item);
    });
    return false;
  };
  return inspect(body) ? "Use local media files and the whisper-py transcription engine." : null;
}

let networkGuardInstalled = false;
export function installLocalNetworkGuard(): void {
  if (!localOnly() || networkGuardInstalled) return;
  networkGuardInstalled = true;
  const fetch = globalThis.fetch;
  globalThis.fetch = (input, init) => {
    const url = new URL(typeof input === "string" ? input : input instanceof URL ? input.href : input.url);
    if (!["127.0.0.1", "localhost", "[::1]"].includes(url.hostname) || !["http:", "https:"].includes(url.protocol)) {
      return Promise.reject(new Error("External network requests are disabled in the local installation."));
    }
    return fetch(input, { ...init, redirect: "error" });
  };
}

export const localMcpTools = new Set([
  "transcribe_podcast", "transcribe_start", "job_status", "suggest_clips",
  "create_clip", "batch_create_clips", "get_ui_state", "modify_clip", "toggle_clip",
  "update_settings", "list_outputs", "set_video", "import_transcript", "parse_transcript",
]);

File: backend/config/policy.py
"""Installation policy shared by backend entry points and provider services."""
import os

STRICT_POLICY = "codex-then-claude"


def local_only() -> bool:
    return os.environ.get("PODCLI_LOCAL_ONLY") == "1"


def strict_ai() -> bool:
    return local_only() or os.environ.get("PODCLI_AI_PROVIDER", "").strip().lower() == STRICT_POLICY


class PolicyError(RuntimeError):
    pass


def require_external(feature: str) -> None:
    if local_only():
        raise PolicyError(f"{feature} is disabled in this local installation.")


def require_cloud() -> None:
    if strict_ai():
        raise PolicyError("Cloud services and direct AI APIs are disabled by the Codex → Claude policy.")


def validate_task(task_type: str, params: dict) -> None:
    if not local_only():
        return
    if task_type in {"manage_integrations", "run_integration_tool"} or (task_type == "manage_config" and params.get("action", "status") != "status"):
        raise PolicyError(f"{task_type} is disabled in this local installation.")
    if task_type == "manage_env" and params.get("action", "list") != "list":
        raise PolicyError("Installation settings are managed in the local profile.")
    if task_type == "transcribe":
        if params.get("engine") not in (None, "", "whisper-py"):
            raise PolicyError("This installation uses local whisper-py transcription.")
        params["engine"] = "whisper-py"
        params["enable_diarization"] = False
        if params.get("model_size", "base") != "base":
            raise PolicyError("Only the installed Whisper base model is enabled.")
    def inspect(value):
        if isinstance(value, dict):
            for key, item in value.items():
                if key in {"url", "video_url", "source_url"} and item:
                    raise PolicyError("Remote media downloads are disabled; select a local file.")
                if key in {"file_path", "video_path", "audio_path", "logo_path", "intro_path", "outro_path", "image_path"} and isinstance(item, str):
                    if "://" in item or item.startswith(("\\\\", "//")):
                        raise PolicyError("Media must use a local filesystem path.")
                inspect(item)
        elif isinstance(value, list):
            for item in value:
                inspect(item)
    inspect(params)


_guard_installed = False


def install_network_guard() -> None:
    global _guard_installed
    if not local_only() or _guard_installed:
        return
    import sys
    def guard(event, args):
        if event == "socket.getaddrinfo":
            host = args[0]
        elif event == "socket.connect" and isinstance(args[1], tuple):
            host = args[1][0]
        else:
            return
        if host not in {"localhost", "127.0.0.1", "::1", b"localhost", b"127.0.0.1", b"::1"}:
            raise PolicyError("External network requests are disabled in the local backend.")
    sys.addaudithook(guard)
    _guard_installed = True
